File: run_inf.cc
// Usage:
//
//  g++ run_inf.cc -o ff_infer \
//      -I/path/to/onnxruntime/include \
//      -L/path/to/onnxruntime/lib -lonnxruntime
//
//  ./ff_infer path/to/model.onnx [path/to/model.json]

#include "onnxruntime_cxx_api.h"
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <stdexcept>
#include <memory>
#include <algorithm>
#include <unordered_map>
#include <map>
#include <optional>
#include <cmath>

#include <nlohmann/json.hpp>

namespace ff_interface {

// Default overridden by JSON
static const std::vector<std::string> kDefaultFeatureOrder = {
    "pt",
    "eta",
    "mass",
    "seedingJet_pt",
    "seedingJet_eta",
    "seedingJet_mass",
    "decayMode_0",
    "decayMode_1",
    "decayMode_2",
    "decayMode_10",
    "decayMode_11",
    "btagPNetB",
    "btagPNetCvB",
    "btagPNetCvL",
    "btagPNetCvNotB",
    "btagPNetQvG"
};

static const std::vector<std::string> kDefaultContFeatureNames = {
    "pt",
    "eta",
    "mass",
    "seedingJet_pt",
    "seedingJet_eta",
    "seedingJet_mass",
    "btagPNetB",
    "btagPNetCvB",
    "btagPNetCvL",
    "btagPNetCvNotB",
    "btagPNetQvG"
};

class FFNetONNXRunner {
public:
    FFNetONNXRunner(const std::string& model_path,
                    const std::string& config_path = std::string())
        : env(ORT_LOGGING_LEVEL_WARNING, "FFNetInference"),
          session(nullptr),
          feature_order(kDefaultFeatureOrder),
          cont_feature_names(kDefaultContFeatureNames)
    {
        check_model_exists_(model_path);
        load_config_(model_path, config_path);
        init_feature_layout_();

        Ort::SessionOptions session_options;
        session_options.SetIntraOpNumThreads(1);
        session_options.SetGraphOptimizationLevel(GraphOptimizationLevel::ORT_ENABLE_EXTENDED);

        session = Ort::Session(env, model_path.c_str(), session_options);

        std::cout << "[info] ONNX model loaded from " << model_path << "\n";
        std::cout << "[info] feature_order size = " << feature_order.size() << "\n";
    }

    float compute_w_ff(const std::vector<float>& cont_features, int decayMode) const {
        if (cont_features.size() != cont_feature_names.size()) {
            std::cerr << "[warn] cont_features.size() = " << cont_features.size()
                      << " (expected " << cont_feature_names.size() << ")\n";
        }

        std::vector<float> cf = cont_features;
        if (cf.size() < cont_feature_names.size()) {
            cf.resize(cont_feature_names.size(), 0.0f);
        }

        auto raw_input = build_input_row_(cf, decayMode);
        return run_onnx_(raw_input);
    }


    float compute_w_ff(const std::map<std::string, float>& features,
                       std::optional<int> decayMode = std::nullopt) const
    {
        int dm = 0;
        if (decayMode.has_value()) {
            dm = decayMode.value();
        } else {
            auto it = features.find("decayMode");
            if (it == features.end()) {
                throw std::runtime_error(
                    "Mapping input must contain key 'decayMode' if no decayMode is passed explicitly.");
            }
            dm = static_cast<int>(std::lround(it->second));
        }

        std::vector<float> cf(cont_feature_names.size(), 0.0f);
        for (std::size_t i = 0; i < cont_feature_names.size(); ++i) {
            const auto& name = cont_feature_names[i];
            auto it = features.find(name);
            if (it != features.end()) {
                cf[i] = it->second;
            }
        }

        auto raw_input = build_input_row_(cf, dm);
        return run_onnx_(raw_input);
    }

private:
    Ort::Env env;
    mutable Ort::Session session;

    std::vector<std::string> feature_order;
    std::vector<std::string> cont_feature_names;

    std::unordered_map<std::string, std::size_t> feature_index;
    std::vector<std::pair<int, std::size_t>> decay_mode_indices; 

    static void check_model_exists_(const std::string& model_path) {
        std::ifstream f(model_path.c_str());
        if (!f.good()) {
            throw std::runtime_error("Could not find ONNX model file: " + model_path);
        }
    }
    static std::string default_config_path_(const std::string& model_path) {
        std::string cfg = model_path;
        const std::string onnx = ".onnx";
        if (cfg.size() >= onnx.size() &&
            cfg.compare(cfg.size() - onnx.size(), onnx.size(), onnx) == 0) {
            cfg.replace(cfg.size() - onnx.size(), onnx.size(), ".json");
        } else {
            cfg += ".json";
        }
        return cfg;
    }

    void load_config_(const std::string& model_path,
                      const std::string& config_path)
    {
        std::string cfg = config_path;
        if (cfg.empty()) {
            cfg = default_config_path_(model_path);
        }

        std::ifstream in(cfg.c_str());
        if (!in.good()) {
            std::cout << "[info] No config at " << cfg
                      << ", using built-in feature_order.\n";
            return;
        }

        std::cout << "[info] Loading input config from " << cfg << "\n";
        nlohmann::json j;
        in >> j;

        if (j.contains("feature_order")) {
            feature_order.clear();
            for (const auto& x : j["feature_order"]) {
                feature_order.push_back(x.get<std::string>());
            }
        }
        if (j.contains("cont_feature_names")) {
            cont_feature_names.clear();
            for (const auto& x : j["cont_feature_names"]) {
                cont_feature_names.push_back(x.get<std::string>());
            }
        }

        std::cout << "[info] Config loaded: feature_order=" << feature_order.size()
                  << ", cont_feature_names=" << cont_feature_names.size() << "\n";
    }

    void init_feature_layout_() {
        feature_index.clear();
        for (std::size_t i = 0; i < feature_order.size(); ++i) {
            feature_index[feature_order[i]] = i;
        }

        decay_mode_indices.clear();
        for (std::size_t i = 0; i < feature_order.size(); ++i) {
            const std::string& name = feature_order[i];
            if (name.rfind("decayMode_", 0) == 0) {
                const std::string suffix = name.substr(std::string("decayMode_").size());
                try {
                    int dm_val = std::stoi(suffix);
                    decay_mode_indices.emplace_back(dm_val, i);
                } catch (const std::exception&) {
                    std::cerr << "[warn] Could not parse decayMode from '" << name << "'\n";
                }
            }
        }

        if (decay_mode_indices.empty()) {
            std::cerr << "[warn] No decayMode_* features found in feature_order\n";
        }
    }

    std::vector<float> build_input_row_(const std::vector<float>& cont_vec,
                                        int decayMode) const
    {
        std::vector<float> raw_input(feature_order.size(), 0.0f);
        for (std::size_t i = 0; i < cont_feature_names.size() && i < cont_vec.size(); ++i) {
            const auto& name = cont_feature_names[i];
            auto it = feature_index.find(name);
            if (it != feature_index.end()) {
                raw_input[it->second] = cont_vec[i];
            }
        }
        if (decay_mode_indices.empty()) {
            std::cerr << "[warn] No decayMode_* entries; decayMode ignored\n";
        } else {
            bool matched = false;
            for (const auto& [dm_val, idx] : decay_mode_indices) {
                if (idx >= raw_input.size()) continue;
                if (dm_val == decayMode) {
                    raw_input[idx] = 1.0f;
                    matched = true;
                } else {
                    raw_input[idx] = 0.0f;
                }
            }
            if (!matched) {
                std::cerr << "[warn] decayMode " << decayMode
                          << " not in configured decayMode_*; one-hot will be all zeros\n";
            }
        }

        return raw_input;
    }

    float run_onnx_(const std::vector<float>& raw_input) const {
        Ort::MemoryInfo memory_info = Ort::MemoryInfo::CreateCpu(OrtArenaAllocator, OrtMemTypeDefault);
        std::vector<int64_t> input_shape = {1, static_cast<int64_t>(raw_input.size())};

        Ort::Value input_tensor = Ort::Value::CreateTensor<float>(
            memory_info,
            const_cast<float*>(raw_input.data()),
            raw_input.size(),
            input_shape.data(),
            input_shape.size()
        );

        const char* input_names[]  = {"raw_input"};
        const char* output_names[] = {"w_ff"};

        auto output_tensors = session.Run(
            Ort::RunOptions{nullptr},
            input_names,  &input_tensor, 1,
            output_names, 1
        );

        float result = output_tensors[0].GetTensorMutableData<float>()[0];
        return result;
    }
};

std::unique_ptr<FFNetONNXRunner> g_ff_runner_instance;

void initialize_ff_runner(const std::string& model_path,
                          const std::string& config_path = std::string()) {
    g_ff_runner_instance = std::make_unique<FFNetONNXRunner>(model_path, config_path);
}

FFNetONNXRunner& get_ff_runner() {
    return *g_ff_runner_instance;
}

} // namespace ff_interface

int main(int argc, char* argv[]) {
    if (argc < 2 || argc > 3) {
        std::cerr << "Usage: " << argv[0] << " model.onnx [model.json]\n";
        return 1;
    }

    const std::string model_path = argv[1];
    const std::string config_path = (argc == 3) ? argv[2] : std::string();

    try {
        ff_interface::initialize_ff_runner(model_path, config_path);
        auto& runner = ff_interface::get_ff_runner();

        // Example 1: mapping with names 
        std::map<std::string, float> features_map = {
            {"pt",            45.0f},
            {"eta",           0.3f},
            {"mass",          1.2f},
            {"seedingJet_pt", 50.0f},
            {"seedingJet_eta",0.1f},
            {"seedingJet_mass",10.0f},
            {"btagPNetB",     0.2f},
            {"btagPNetCvB",   0.1f},
            {"btagPNetCvL",   0.4f},
            {"btagPNetCvNotB",0.3f},
            {"btagPNetQvG",   0.5f},
            {"decayMode",     1.0f}
        };

        float w_map = runner.compute_w_ff(features_map);

        // Example 2: plain vector 
        std::vector<float> cont_features = {
            45.0f,  // pt
            0.3f,   // eta
            1.2f,   // mass
            50.0f,  // seedingJet_pt
            0.1f,   // seedingJet_eta
            10.0f,  // seedingJet_mass
            0.2f,   // btagPNetB
            0.1f,   // btagPNetCvB
            0.4f,   // btagPNetCvL
            0.3f,   // btagPNetCvNotB
            0.5f    // btagPNetQvG
        };
        int decayMode = 1;

        float w_vec = runner.compute_w_ff(cont_features, decayMode);

        std::cout << "\n=== SUMMARY ===\n";
        std::cout << "Model path:      " << model_path << "\n";
        if (!config_path.empty())
            std::cout << "Config path:     " << config_path << "\n";
        std::cout << "w_ff (mapping):  " << w_map << "\n";
        std::cout << "w_ff (vector):   " << w_vec << "\n";
        return 0;
    } catch (const std::exception& ex) {
        std::cerr << "[error] " << ex.what() << "\n";
        return 1;
    }
}

File: run_inf.py
#!/usr/bin/env python3
"""
Usage:
  python run_inf.py --onnx models/model_v1911.onnx [--config models/model_v1911.json]

To import FFNetONNXRunner in analysis:

  from run_ff import FFNetONNXRunner
  runner = FFNetONNXRunner("model.onnx", "model.json")
  w = runner.compute_w_ff(features)
"""

import argparse
import json
import logging
import os
from pathlib import Path
from typing import Mapping, Optional, Sequence, Union, List

import numpy as np
import onnxruntime as ort

LOGGER = logging.getLogger("FF_ONNX")

# Default ONNX feature order overridden by inout JSON config
DEFAULT_FEATURE_ORDER: List[str] = [
    "pt",
    "eta",
    "mass",
    "seedingJet_pt",
    "seedingJet_eta",
    "seedingJet_mass",
    "decayMode_0",
    "decayMode_1",
    "decayMode_2",
    "decayMode_10",
    "decayMode_11",
    "btagPNetB",
    "btagPNetCvB",
    "btagPNetCvL",
    "btagPNetCvNotB",
    "btagPNetQvG",
]

DEFAULT_CONT_FEATURE_NAMES: List[str] = [
    "pt",
    "eta",
    "mass",
    "seedingJet_pt",
    "seedingJet_eta",
    "seedingJet_mass",
    "btagPNetB",
    "btagPNetCvB",
    "btagPNetCvL",
    "btagPNetCvNotB",
    "btagPNetQvG",
]

ARRAY_WITH_DM_SIZE = len(DEFAULT_CONT_FEATURE_NAMES) + 1
ARRAY_DM_INDEX = 6  

def setup_logging(verbose: bool) -> None:
    level = logging.DEBUG if verbose else logging.INFO
    logging.basicConfig(
        level=level, format="%(asctime)s [%(levelname)s] %(message)s"
    )

class FFNetONNXRunner:
    def __init__(self, model_path: str, config_path: Optional[str] = None):
        if not os.path.exists(model_path):
            raise FileNotFoundError(f"Could not find ONNX model file: {model_path}")

        self.feature_order = list(DEFAULT_FEATURE_ORDER)
        self.cont_feature_names = list(DEFAULT_CONT_FEATURE_NAMES)
        self.decay_mode_indices: List[tuple[int, int]] = []
        self.feature_index: dict[str, int] = {}

        self._load_config(model_path, config_path)
        self._init_feature_layout()

        LOGGER.info("Loading ONNX model from %s", model_path)
        self.session = ort.InferenceSession(model_path, providers=["CPUExecutionProvider"])
        self.input_name = "raw_input"
        self.output_name = "w_ff"

        LOGGER.info("ONNX session ready (features: %d)", len(self.feature_order))
        LOGGER.debug("feature_order: %s", self.feature_order)
        LOGGER.debug("cont_feature_names: %s", self.cont_feature_names)

    def _load_config(self, model_path: str, config_path: Optional[str]) -> None:
        if config_path:
            cfg_path = Path(config_path)
        else:
            cfg_path = Path(model_path).with_suffix(".json")

        if not cfg_path.exists():
            LOGGER.info("No config found at %s, using defaults.", cfg_path)
            return

        LOGGER.info("Loading input config from %s", cfg_path)
        data = json.loads(cfg_path.read_text())

        if not isinstance(data, dict):
            raise RuntimeError(f"Config file {cfg_path} must contain a JSON object.")

        if "feature_order" in data:
            self.feature_order = [str(x) for x in data["feature_order"]]

        if "cont_feature_names" in data:
            self.cont_feature_names = [str(x) for x in data["cont_feature_names"]]

        LOGGER.info(
            "Config loaded: feature_order=%d, cont_feature_names=%d",
            len(self.feature_order),
            len(self.cont_feature_names),
        )

    def _init_feature_layout(self) -> None:
        self.feature_index = {name: i for i, name in enumerate(self.feature_order)}
        self.decay_mode_indices = []
        for idx, name in enumerate(self.feature_order):
            if name.startswith("decayMode_"):
                try:
                    dm_val = int(name.split("_", 1)[1])
                    self.decay_mode_indices.append((dm_val, idx))
                except ValueError:
                    LOGGER.warning("Could not parse decayMode from '%s'", name)

        if not self.decay_mode_indices:
            LOGGER.warning("No decayMode_* entries found in feature_order")
        else:
            LOGGER.debug("decayMode layout: %s", self.decay_mode_indices)
        missing = [n for n in self.cont_feature_names if n not in self.feature_index]
        if missing:
            LOGGER.warning(
                "Some cont_feature_names not present in feature_order: %s", missing
            )

    def _from_mapping(
        self,
        features: Mapping[str, float],
        decay_mode: Optional[int],
    ) -> tuple[np.ndarray, int]:
        if decay_mode is None:
            if "decayMode" not in features:
                raise ValueError(
                    "Mapping input must contain 'decayMode' if no decay_mode argument is given."
                )
            dm = int(features["decayMode"])
        else:
            dm = int(decay_mode)

        cf = np.zeros(len(self.cont_feature_names), dtype=np.float32)
        for i, name in enumerate(self.cont_feature_names):
            if name in features:
                cf[i] = float(features[name])
        return cf, dm

    def _from_array(
        self,
        arr_like: Union[Sequence[float], np.ndarray],
        decay_mode: Optional[int],
    ) -> tuple[np.ndarray, int]:
        arr = np.asarray(arr_like, dtype=np.float32).ravel()
        if decay_mode is None:
            if arr.size != ARRAY_WITH_DM_SIZE:
                raise ValueError(
                    f"Array input (without explicit decay_mode) must have length "
                    f"{ARRAY_WITH_DM_SIZE} (cont + decayMode), got {arr.size}."
                )
            dm = int(round(float(arr[ARRAY_DM_INDEX])))

            cf = np.zeros(len(self.cont_feature_names), dtype=np.float32)
            cf[0:6] = arr[0:6]
            cf[6:] = arr[ARRAY_DM_INDEX + 1 :]
        else:
            dm = int(decay_mode)
            if arr.size != len(self.cont_feature_names):
                raise ValueError(
                    f"Array input (with explicit decay_mode) must have length "
                    f"{len(self.cont_feature_names)}, got {arr.size}."
                )
            cf = arr.copy()

        return cf, dm

    def _build_input_tensor(self, cf: np.ndarray, decay_mode: int) -> np.ndarray:
        x = np.zeros((1, len(self.feature_order)), dtype=np.float32)
        for i, name in enumerate(self.cont_feature_names):
            idx = self.feature_index.get(name)
            if idx is not None:
                x[0, idx] = cf[i]
        if not self.decay_mode_indices:
            LOGGER.warning("No decayMode_* features in feature_order, decayMode ignored.")
        else:
            matched = False
            for dm_val, idx in self.decay_mode_indices:
                if idx >= x.shape[1]:
                    continue
                x[0, idx] = 1.0 if dm_val == decay_mode else 0.0
                if dm_val == decay_mode:
                    matched = True
            if not matched:
                LOGGER.warning(
                    "DecayMode %d not among configured decayMode_* values; one-hot will be all zeros.",
                    decay_mode,
                )

        LOGGER.debug("Input tensor row: %s", x[0])
        return x

    def compute_w_ff(
        self,
        features: Union[Mapping[str, float], Sequence[float], np.ndarray],
        decay_mode: Optional[int] = None,
    ) -> float:
        if isinstance(features, Mapping):
            cf, dm = self._from_mapping(features, decay_mode)
        else:
            cf, dm = self._from_array(features, decay_mode)

        LOGGER.info("Running inference (decayMode=%d)", dm)
        x = self._build_input_tensor(cf, dm)

        out = self.session.run([self.output_name], {self.input_name: x})[0]
        out_flat = np.asarray(out, dtype=np.float32).ravel()
        if out_flat.size == 0:
            raise RuntimeError("Model output is empty.")
        return float(out_flat[0])


_runner_instance: Optional[FFNetONNXRunner] = None

def initialize_ff_runner(model_path: str, config_path: Optional[str] = None) -> None:
    global _runner_instance
    _runner_instance = FFNetONNXRunner(model_path, config_path)

def get_ff_runner() -> FFNetONNXRunner:
    if _runner_instance is None:
        raise RuntimeError("FFNetONNXRunner has not been initialized.")
    return _runner_instance

def main() -> None:
    ap = argparse.ArgumentParser("FF ONNX inference")
    ap.add_argument("--onnx", required=True, help="Path to model.onnx")
    ap.add_argument("--config", help="Optional JSON config with feature_order")
    ap.add_argument("-v", "--verbose", action="store_true", help="Enable DEBUG logs")
    args = ap.parse_args()

    setup_logging(args.verbose)

    runner = FFNetONNXRunner(args.onnx, args.config)

    # Example 1: dict with names
    features_dict = {
        "pt": 45.0,
        "eta": 0.3,
        "mass": 1.2,
        "seedingJet_pt": 50.0,
        "seedingJet_eta": 0.1,
        "seedingJet_mass": 10.0,
        "btagPNetB": 0.2,
        "btagPNetCvB": 0.1,
        "btagPNetCvL": 0.4,
        "btagPNetCvNotB": 0.3,
        "btagPNetQvG": 0.5,
        "decayMode": 1,
    }
    w_dict = runner.compute_w_ff(features_dict)

    # Example 2: plain numpy array 
    features_arr = np.array(
        [
            45.0,  # pt
            0.3,   # eta
            1.2,   # mass
            50.0,  # seedingJet_pt
            0.1,   # seedingJet_eta
            10.0,  # seedingJet_mass
            1.0,   # decayMode 
            0.2,   # btagPNetB
            0.1,   # btagPNetCvB
            0.4,   # btagPNetCvL
            0.3,   # btagPNetCvNotB
            0.5,   # btagPNetQvG
        ],
        dtype=np.float32,
    )
    w_arr = runner.compute_w_ff(features_arr)

    print("\n=== SUMMARY ===")
    print(f"ONNX model:   {args.onnx}")
    if args.config:
        print(f"Config:       {args.config}")
    print(f"feature_order length: {len(runner.feature_order)}")
    print(f"w_ff (dict):  {w_dict:.6g}")
    print(f"w_ff (array): {w_arr:.6g}")


if __name__ == "__main__":
    main()